### PR TITLE
[RFR][1LP] Optimized Quickstart

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this enables `python -m cfme.scripting.quickstart` to set up the environment, configuration file as well as the base package installs on redhat based systems

it tries to avoid recreating the virtualenv based on path existence
it tries to avoid reinstalling the package set based on the sha1 of the frozen packages
it tries to help with linking config-files from the yaml repo and reminds of encryption

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit d3277ecf50fe7abc1c9c3f7bdcd21fa99d1e1df6:

`cfme/__init__.py`:
- :bangbang: Line 2:5: **E265** *block comment should start with '# '*
- :bangbang: Line 11:5: **E303** *too many blank lines (2)*
- :bangbang: Line 43:5: **E303** *too many blank lines (2)*

`cfme/scripting/disable_bytecode.py`:
- :grey_exclamation: Line 1:1: **T001** *print statement found.*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/scripting/quickstart.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+REQUIREMENT_FILE = 'requirements/frozen.txt'
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--mk_virtualenv", default="../cfme_venv")
+parser.add_argument("--system-site-packages", action="store_true")
+parser.add_argument("--config_path", default="../cfme-qe-yamls/complete/")
+
+
+def call_or_exit(*k, **kw):
```
Comment (izapolsk) [inline cfme/scripting/quickstart.py]: it looks very similar to check_call

File: cfme/scripting/quickstart.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    venv_call(venv_path, 'pip', 'install', '-e', '.')
+
+
+def disable_bytecode(venv_path):
+    venv_call(venv_path, 'python', '-m', 'cfme.scripting.disable_bytecode')
+
+
+def link_config_files(src, dest):
+    if not os.path.isdir(src):
+        print("WARNING: not linking config files,", src, "missig")
```
Comment (izapolsk) [inline cfme/scripting/quickstart.py]: typo

File: cfme/scripting/quickstart.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
+from __future__ import print_function
+
+import os
+import sys
+import argparse
+import subprocess
+import hashlib
+
+IS_SCRIPT = sys.argv[0] == __file__
```
Comment (izapolsk) [inline cfme/scripting/quickstart.py]: :+1:

File: cfme/scripting/quickstart.py (hunk 1 above)
Comment (RonnyPfannschmidt) [inline cfme/scripting/quickstart.py]: yes, but checked_call prints a long traceback when used

File: cfme/scripting/quickstart.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+CWD = os.getcwd()  # we expect to be in the workdir
+REDEHAT_BASED = os.path.isfile('/etc/redhat-release')
```
Comment (RonnyPfannschmidt) [inline cfme/scripting/quickstart.py]: oh man, thats one of those you never see yourself

File: cfme/scripting/quickstart.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--mk_virtualenv", default="../cfme_venv")
+parser.add_argument("--system-site-packages", action="store_true")
+parser.add_argument("--config_path", default="../cfme-qe-yamls/complete/")
+
+
+def call_or_exit(command, shell=False):
+    try:
+        res = subprocess.call(command, shell=shell)
```
Comment (mfalesni) [inline cfme/scripting/quickstart.py]: Maybe add `print('Executing: {}'.format(' '.join(command)))` so the person who executes it sees exactly what happens.

Comment (RonnyPfannschmidt) [inline cfme/scripting/quickstart.py]: good idea

**Comment by dajoRH** [comment]
Lint report for commit 5ddd559a20e1fd4433f8edb1d9c2cda36ae90f30:

`cfme/scripting/disable_bytecode.py`:
- :bangbang: Line 22:1: **E305** *expected 2 blank lines after class or function definition, found 1*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: docs/getting_started.rst [hunk 6]
```diff
@@ -1,116 +1,43 @@
...
 setup and execution scripts.
 
+
+Obtaining what you need (Project Setup)
+----------------------------------------
+
+* create a dedicated folder for working with the integration tests,
+  our automated quickstart has expectations that best work out
+  when it has its own place
+* obtain the ``cfme_tests`` repository (via ``git clone git@github.com:ManageIQ/integration_tests.git cfme_tests``)
```
Comment (mfalesni) [inline docs/getting_started.rst]: Wrong. You should say they need to fork if they want to contribute.

File: cfme/scripting/quickstart.py (hunk 5 above)
Comment (mfalesni) [inline cfme/scripting/quickstart.py]: You haven't added this suggestion, is that intended?

File: cfme/scripting/quickstart.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+
+def install_system_packages():
+    if REDHAT_BASED:
+        call_or_exit(
+            "pkcon install -yq"
+            " python-virtualenv gcc postgresql-devel libxml2-devel"
+            " libxslt-devel zeromq3-devel libcurl-devel"
+            " redhat-rpm-config gcc-c++ openssl-devel"
+            " libffi-devel python-devel", shell=True)
```
Comment (mfalesni) [inline cfme/scripting/quickstart.py]: there should be an else to tell the user to install required packages on their distro for correct setup.

File: cfme/scripting/quickstart.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+
+def hash_file(path):
+    content_hash = hashlib.sha1()
+    with open(path, 'rb') as fp:
+        content_hash.update(fp.read())
+    return content_hash.hexdigest()
+
+
+def install_requirements(venv_path, quiet=False):
```
Comment (psav) [inline cfme/scripting/quickstart.py]: This install mechanism doesn't seem to take the PYCURL_SSL_LIBRARY env var into account?

File: docs/getting_started.rst [hunk 9]
```diff
@@ -1,116 +1,44 @@
...
-  repository. Required YAML files are ``env``, ``cfme_data``, ``credentials``. If the file's
-  extension is ``.yaml`` it is loaded normally, if its extension is ``.eyaml`` then it is encrypted
-  and you need to have the decryption key in the ``cfme_tests/`` directory. You can also start them
-  from scratch by copying the templates in ``conf/`` and editing them to suit the environment you
-  use.
-* Set up a local selenium server that opens browser windows somewhere other than your
-  desktop. There is a Docker based solution for the browser, look at the script
-  ``scripts/dockerbot/sel_container.py``. That ensures you have the proper versions of browsers. You
-  can also set everything up in your system using Xvnc - :doc:`guides/vnc_selenium` .
 * Test! Run py.test. (This takes a long time, Ctrl-C will stop it)
```
Comment (psav) [inline docs/getting_started.rst]: I thought there was supposed to be a wrapper component to this that we run the py.test by passing the args to the wrapper.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/scripting/quickstart.py (hunk 8 above)
Comment (RonnyPfannschmidt) [inline cfme/scripting/quickstart.py]: now taken

**Comment by dajoRH** [comment]
Lint report for commit 7d7142655a625c45a05a09c87312e03d2b0fc682:

`cfme/__init__.py`:
- :bangbang: Line 2:5: **E265** *block comment should start with '# '*
- :bangbang: Line 10:5: **E303** *too many blank lines (2)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a70f370ce1b2f1a67659e494a840bb0d969e905b:

`cfme/__init__.py`:
- :bangbang: Line 2:5: **E265** *block comment should start with '# '*
- :bangbang: Line 10:5: **E303** *too many blank lines (2)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e4e24ef18808c5cabfd3c5482d812cff5123482f:

`cfme/__init__.py`:
- :bangbang: Line 2:5: **E265** *block comment should start with '# '*
- :bangbang: Line 10:5: **E303** *too many blank lines (2)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by psav** [comment]
Issues to resole with this branch
- [ ] pycurl does not compile with the right backend
- [ ] ```xfai``` is present in the tests, it should be ```xfail```
- [ ] Invoking sudo yum each time is needlessly time consuming, we could use ```rpm -q <package>``` for each package and construct a string of missing dependencies to do a yum install, why? rpm checking for a single package is quicker, and it doesn't need sudo
- [ ] Is it the intention that multiple setups of this would use the same venv_location, if I ran this in two separate directories of the same parent?
- [ ] Some warning should be given that a) you need to run the source for the venv, and b) that its location is ```<here>```

**Comment by dajoRH** [comment]
Lint report for commit 72feef70ec93af1ac4b1261a6c23b978d1d439cb:

`cfme/__init__.py`:
- :bangbang: Line 2:5: **E265** *block comment should start with '# '*
- :bangbang: Line 10:5: **E303** *too many blank lines (2)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by RonnyPfannschmidt** [comment]
the package manager is now only invoked by externalized quick-start invocation, on active virtualenv we assume we dont need to

custom virtualenv locations per custom checkout can be ensured via 'python -m cfme.scripting.quickstart --mk_virtualenv ../mycustomenv' else the default is assumed

File: cfme/scripting/quickstart.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        INSTALL_COMMAND = 'yum install -y'
+    if not IS_ROOT:
+        INSTALL_COMMAND = 'sudo ' + INSTALL_COMMAND
+
+
+def command_text(command, shell):
+    if shell:
+        return command
+    else:
+        command = [repr(x) if ' ' in x else x for x in command]
```
Comment (mfalesni) [inline cfme/scripting/quickstart.py]: Maybe `utils/quote.py::quote` ?

Comment (RonnyPfannschmidt) [inline cfme/scripting/quickstart.py]: @mfalesni utils is not importable because we put all the things into a toplevel  __init__

**Comment by dajoRH** [comment]
Lint report for commit 4f2a7f3d35d67415a83ad723afcfba09c4810957:

`cfme/scripting/quickstart.py`:
- :grey_exclamation: Line 180:35: **T101** *Python 2.x reserved word print used.*
- :grey_exclamation: Line 183:52: **T101** *Python 2.x reserved word print used.*
- :grey_exclamation: Line 195:34: **T101** *Python 2.x reserved word print used.*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: setup.py [hunk 11]
```diff
@@ -45,24 +46,6 @@
             'kubernetes = cfme.containers.provider.kubernetes:KubernetesProvider',
             'openshift = cfme.containers.provider.openshift:OpenshiftProvider',
         ],
-        'manageiq.vm_categories':
```
Comment (psav) [inline setup.py]: **REQUIRED:** These shouldn't be removed afaik

Comment (RonnyPfannschmidt) [inline setup.py]: good find, i missed this rebase artifact

File: docs/getting_started.rst [hunk 12]
```diff
@@ -223,15 +151,15 @@ started as quickly as possible:
 
 * If you want to test against multiple appliances, use the ``--appliance w.x.y.z`` parameter. Eg. if
   you have appliances ``1.2.3.4`` and ``2.3.4.5``, then append ``--appliance 1.2.3.4 --appliance 2.3.4.5``
-  to the ``py.test`` command. Due to a glitch that has not been resolved yet, you should set the
+  to the ``cfme-runtest`` command. Due to a glitch that has not been resolved yet, you should set the
```
Comment (psav) [inline docs/getting_started.rst]: **REQUIRED:** This should be miq-runtest

**Comment by dajoRH** [comment]
Lint report for commit d6a099e55652e78f35b817f2765567e7afefb716:

`cfme/scripting/quickstart.py`:
- :grey_exclamation: Line 180:35: **T101** *Python 2.x reserved word print used.*
- :grey_exclamation: Line 183:52: **T101** *Python 2.x reserved word print used.*
- :grey_exclamation: Line 195:34: **T101** *Python 2.x reserved word print used.*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: docs/getting_started.rst [hunk 13]
```diff
@@ -1,118 +1,46 @@
...
+Obtaining what you need (Project Setup)
+----------------------------------------
+
+* create a dedicated folder for working with the integration tests,
+  our automated quickstart has expectations that best work out
+  when it has its own place
+* obtain the ``cfme_tests`` repository
+  (`fork and clone <https://github.com/ManageIQ/integration_tests/fork>`_)
+* if you have access, obtain the ``cfme-qe-yamls`` repository (internal url, please ask the team) and
+  put the ``.yaml_key`` file into the root of the `cfme_tests`` repository.
```
Comment (mfalesni) [inline docs/getting_started.rst]: Bad syntax of the backticks, they need to be 2

File: cfme/scripting/quickstart.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+except ImportError:
+    sys.path.insert(0, os.path.join(
+        os.path.dirname(os.path.dirname(
+            os.path.dirname(__file__))),
+        'requirements', 'quickstart',
+        'distro-1.0.4-py2.py3-none-any.whl'))
+    import distro
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--mk_virtualenv", default="../cfme_venv")
```
Comment (psav) [inline cfme/scripting/quickstart.py]: **REQUIRED:** Something I only just noticed on running again, can we have consistency between ```_``` and ```-```

**Comment by dajoRH** [comment]
Lint report for commit dd06357016a906811cc8f0296b97925cc96f705b:

`cfme/scripting/quickstart.py`:
- :grey_exclamation: Line 180:35: **T101** *Python 2.x reserved word print used.*
- :grey_exclamation: Line 183:52: **T101** *Python 2.x reserved word print used.*
- :grey_exclamation: Line 195:34: **T101** *Python 2.x reserved word print used.*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Pull request merged**